2022-MAY/answer.md

```diff
@@ -0,0 +1,163 @@
+JOIN Department d ON b.DeptID = d.DeptID
+```
+
+### Question 2
+
+a)
+
+![ERD Diagram](./question_2a.drawio.png)
+
+b)
+
+CUSTOMER(<ins>CustomerID</ins>, CustomerName, ContactNo)
+
+CAKE(<ins>CakeID</ins>, Description, Price)
+
+INGREDIENT(<ins>IngredientID</ins>, Name)
+
+ORDER(<ins>OrderID</ins>, CustomerID\*, CakeID\*, OrderDate, Quantity)
+
+CAKE_INGREDIENT(<ins>CakeID\*</ins>, <ins>IngredientID\*</ins>, IngredientAmount)
+
+
+
+### Question 3
+
+a) 
+
+- Insertion Anomaly: Insertion of new customer data requires to insert redundant home and staff data.
+- Modification Anomaly: Modification of price per night data of H01 needs to be made to all records that contains the specific home ID H01 to avoid data inconsistency.
+- Deletion Anomaly: Deletion of H04 will cause the customer C002 record to be removed from the database.
+
+b)
+
+**1NF**
+
+HOME(<ins>HomeID</ins>, Description, PricePerNight, StaffID, StaffHP, <ins>CustomerID</ins>, CustomerName, CustomerHP, CheckInDate, CheckOutDate)
+
+> Use the **entity name of the non-repeating group** as the table name
+
+**2NF**
+
+HOME(<ins>HomeID</ins>, Description, PricePerNight, StaffID, StaffHP)
+
+BOOKING(<ins>HomeID\*</ins>, <ins>CustomerID\*</ins>, <ins>CheckInDate\*</ins>, <ins>CheckOutDate\*</ins>)
+
+CUSTOMER(<ins>CustomerID</ins>, CustomerName, CustomerHP)
+
+> Break it down into **3 table**, which is one for **non-repeating group**, one for **repeating group** and one for **bridge entity**.
+
+**3NF**
+
+HOME(<ins>HomeID</ins>, Description, PricePerNight, StaffID\*)
+
+STAFF(<ins>StaffID</ins>, StaffHP)
+
+BOOKING(<ins>HomeID\*</ins>, <ins>CustomerID\*</ins>, <ins>CheckInDate\*</ins>, <ins>CheckOutDate\*</ins>)
+
+CUSTOMER(<ins>CustomerID</ins>, CustomerName, CustomerHP)
+
+### Question 4
+
+a) 
+
+```sql
+CREATE TABLE Appointment (
+	AppointmentNo VARCHAR(7) NOT NULL,
+	ICNo VARCHAR(20) NOT NULL,
+	VaccineNo VARCHAR(7) NOT NULL,
+	Appointment_Date DATE NOT NULL,
+	Status VARCHAR(30) NOT NULL,
+  Remark VARCHAR(50),
+	PRIMARY KEY (AppointmentNo),
+	FOREIGN KEY (IcNo) REFERENCES Person(IcNo),
+	FOREIGN KEY (VaccineNo) REFERENCES Merchant(VaccineNo)
+  CONSTRAINT check_status CHECK (Status IN ('PENDING', 'COMPLETED', 'CANCELLED'))
+);
+```
+
+c)
+
+```sql
+UPDATE Vaccine
+SET Vaccine_Description = 'Sinovac Dose 1'
+WHERE VaccineNo = 'SIND01';
+```
+
+b)
+
+```sql
+INSERT INTO Merchant VALUES ('600125-12-1234', 'Kimberly Lee', '012-3456789', 'Sabah');
+```
+
+d)
+
+```sql
+DELETE FROM Appointment
+WHERE Appointment_Date = '25-DEC-2021';
+```
+
+e)
+
+```sql
+SELECT P.ICno, P.Name, P.Contact_No
+FROM Person P
+JOIN Appointment a ON P.ICNo = A.ICNo
+WHERE A.status = 'CANCELLED';
+
+```
+
+f)
+
+```sql
+SELECT V.VaccineNo 'Vaccine Number', V.Vaccine_Description 'Vaccine Description', COUNT(A.AppointmentNo) 'No_Of_Doses_Given_2021'
+FROM Vaccine V
+JOIN Appointment A ON V.VaccineNo = A.VaccineNo
+WHERE (A.Appointment_Date BETWEEN '01-JAN-2021' AND '31-DEC-2021') AND (A.Status = 'COMPLETED'); 
```

Comment:
Mana `GROUP BY`

Can use LEFT JOIN instead cuz can include all Vaccines, and COUNT will ignore NULL value if the data from right table is empty